File: src/main/variations/RandomVariation.java
package main.variations;

import main.buttons.Element;
import main.variations.appearances.Appearance;
import main.variations.appearances.Texture;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.math3.distribution.EnumeratedDistribution;
import org.apache.commons.math3.util.Pair;
import processing.core.PImage;
import processing.data.JSONArray;
import processing.data.JSONObject;

import java.util.ArrayList;

public class RandomVariation extends Variation {

    private EnumeratedDistribution<Appearance> random;

    RandomVariation(JSONObject json, Element element) {
        super(json, element);
    }

    @Override
    public void loadImages() {
        JSONArray array = this.json.getJSONArray("textures");
        ArrayList<Pair<Appearance, Double>> list = new ArrayList<>();
        double remainingWeight = 1;
        for (int i = 0; i < array.size(); i++) {
            JSONObject object = array.getJSONObject(i);
            list.add(new Pair<>(Appearance.getAppearance(this, object), object.getDouble("weight")));
            remainingWeight -= object.getDouble("weight");
        }
        list.add(new Pair<>(new Texture(this), remainingWeight)); //chance of getting the original image
        this.random = new EnumeratedDistribution<>(list);
    }

    @Override
    public Appearance getAppearance() {
        return this.random.sample();
    }

//    //even though the null pair indicates we want the original image, it will actually have the fallback image, hence why we need to check and change it
//    @Override
//    public PImage getImage() {
//        PImage image = super.getImage();
//        if (image == Button.error) {
//            return null;
//        } else {
//            return image;
//        }
//    }

    @Override
    public ArrayList<ImmutablePair<PImage, String>> getPairs() {
        ArrayList<ImmutablePair<PImage, String>> list = new ArrayList<>();
        for (Pair<Appearance, Double> pair : this.random.getPmf()) {
            list.addAll(pair.getFirst().getPairs());
        }
        return list;
    }

    //this must be used so it doesn't get a random name every time, but the name corresponding the chosen image
    public String getName(PImage image) {
        for (Pair<Appearance, Double> pair : this.random.getPmf()) {
            if (pair.getFirst().getImage() != null && pair.getFirst().getImage() == image) {
                return pair.getFirst().getName();
            }
        }
        return null;
    }

}

File: src/main/variations/RandomVariation.kt
package main.variations

import main.Element
import main.lateVal
import main.variations.appearances.Appearance
import main.variations.appearances.Appearance.Companion.getAppearance
import main.variations.appearances.ElementTexture
import processing.core.PImage
import processing.data.JSONObject
import kotlin.random.Random

class RandomVariation internal constructor(json: JSONObject, element: Element) : Variation(json, element) {

    private var appearances: List<Appearance> by lateVal()
    private var random: WeightedRandom<Appearance> by lateVal()
    private lateinit var current: Appearance

    override fun loadImages() {
        val array = json.getJSONArray("textures")
        val map = mutableMapOf<Appearance, Double>()
        var remainingWeight = 1.0
        for (i in 0 until array.size()) {
            val `object` = array.getJSONObject(i)
            map[getAppearance(this, `object`)] = `object`.getDouble("weight")
            remainingWeight -= `object`.getDouble("weight")
        }
        //TODO if remaining weight <= 0, means there is no chance for original image
        map[ElementTexture(element)] = remainingWeight //chance of getting the original image
        appearances = map.keys.toList()
        random = WeightedRandom(map)
        current = appearances[0]
    }

    override fun getAppearance(): Appearance {
        return current
    }

    override fun getPairs(): List<Pair<PImage, String>> {
        return appearances.map { it.getPairs() }.flatten()
    }

    fun setCurrentImage() {
        current = random.get()
    }

    //this must be used so it doesn't get a random name every time, but the name corresponding the chosen image
//    fun getName(image: PImage): String? {
//        for (pair in random.pmf) {
//            if (pair.first.getImage() === image) {
//                return pair.first.getName()
//            }
//        }
//        return null
//    }
}

class WeightedRandom<T>(map: Map<T, Number>) {

    private val list = mutableListOf<Triple<Double, Double, T>>()

    init {
        var i = 0.0
        for ((key, value) in map) {
            list.add(Triple(i, i + value.toDouble(), key))
            i += value.toDouble()
        }
    }

    fun get(): T {
        val random = Random.nextDouble(list.last().second)
        for (triple in list) {
            if (random >= triple.first && random < triple.second) {
                return triple.third
            }
        }
        return list.first().third
    }

}